File: src/services/GraphService.ts
import { ServiceKey, ServiceScope } from "@microsoft/sp-core-library";
import { MSGraphClientFactory, MSGraphClientV3,  } from "@microsoft/sp-http";
import { ISite } from "../model/ISite";
import { IApp } from "../model/IApp";

export default class GraphService {
	private msGraphClientFactory: MSGraphClientFactory;
  private client: MSGraphClientV3;

  public static readonly serviceKey: ServiceKey<GraphService> =
    ServiceKey.create<GraphService>('react-application-nav-graph', GraphService);

  constructor(serviceScope: ServiceScope) {  
    serviceScope.whenFinished(async () => {
      this.msGraphClientFactory = serviceScope.consume(MSGraphClientFactory.serviceKey);      
    });
  }

  public async searchSites(queryText: string, start: number): Promise<ISite[]> {
    const searchRawSites = await this.searchRawSites(queryText, start);
    const sites = await this.transformSearchSites(searchRawSites);
    return sites;
  }

  public async servicePrincipals(): Promise<IApp[]> {
    this.client = await this.msGraphClientFactory.getClient('3');
    const response = await this.client
            .api(`servicePrincipals`)
            .version('v1.0')
            .filter(`startswith(DisplayName,'dlgScope')`)  // Assumption
            .get();
    
    const apps: IApp[] = [];
    response.value.forEach((app: any) => {
      apps.push({Id: app.appId, DisplayName: app.displayName });
    });
    return apps;
  }

  public async servicePrincipal(appId: string): Promise<string> {
    this.client = await this.msGraphClientFactory.getClient('3');
    const response = await this.client
            .api(`servicePrincipals`)
            .version('v1.0')
            .filter(`appId eq '${appId}'`)
            .get();
    return response.value[0].displayName;
  }

  public async isSiteAdmin(userEMail: string, currentSiteId: string): Promise<boolean> {
    this.client = await this.msGraphClientFactory.getClient('3');
    const response = await this.client
            .api(`sites/${currentSiteId}/lists/User Information List/items`)
            .version('v1.0')
            .header('Prefer','HonorNonIndexedQueriesWarningMayFailRandomly')
            .expand('fields($select=EMail,IsSiteAdmin)')
            .filter(`fields/EMail eq '${userEMail}'`)
            .get();
    return response.value[0].fields.IsSiteAdmin;
  }

  public async grantPermissions(role: string, appId: string, displayName: string, siteId: string): Promise<any[]> {
    this.client = await this.msGraphClientFactory.getClient('3');
    const requestBody = {
      roles: [
        role
      ],
      grantedToIdentities: [
        {
          application: {
            id: appId,
            displayName: displayName
          }
        }
      ]
    };

    const response = await this.client
            .api(`/sites/${siteId}/permissions`)
            .version('v1.0')    
            .post(requestBody);
    return response;
  }

  private async searchRawSites(queryText: string, start: number): Promise<any[]> {
    this.client = await this.msGraphClientFactory.getClient('3');
    const requestBody = {
      requests: [
          {
              entityTypes: [
                  "site"
              ],
              query: {
                  "queryString": `${queryText}`
              }
          }
      ]
    };

    const response = await this.client
            .api(`search/query`)
            .version('v1.0')
            .skip(start)
            .top(20)   // Limit in batching!      
            .post(requestBody);
    if (response.value[0].hitsContainers[0].total > 0) {
      return response.value[0].hitsContainers[0].hits;
    }
    else return [];
  }

  private transformSearchSites(response: any[]): ISite[] {    
    const items: Array<ISite> = new Array<ISite>();
    if (response !== null && response.length > 0) {
      response.forEach((r: any) => {          
        items.push({ Title: r.resource.displayName, Url: r.resource.webUrl, Id: r.resource.id });
      });
      return items;
    }
    else {
      return [];
    }
  }
}

File: Functions/ApplyPermissions/Services/GraphService.cs
using ApplyPermissions.Model;
using Azure;
using Azure.Core;
using Azure.Identity;
using Google.Protobuf.WellKnownTypes;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using Microsoft.Graph;
using Microsoft.Graph.Models;
using Microsoft.Identity.Client;
using static Microsoft.Graph.CoreConstants;

namespace ApplyPermissions.Services
{
  public class GraphService : IGraphService
  {
    private readonly IConfiguration _config;
    private readonly ILogger _logger;
    private GraphServiceClient? _appGraphClient;

    public GraphService(IConfiguration config, ILogger<GraphService> logger)
    {
      _config = config;
      _logger = logger;
    }

    public async Task<bool> ApplySitePermission(string userAssertion, string siteUrl, string role, string appId)
    {
      _appGraphClient = GetUserGraphClient(userAssertion);
      Uri uri = new Uri(siteUrl);
      string domain = uri.Host;
      var path = uri.LocalPath;
      var site = await _appGraphClient.Sites[$"{domain}:{path}"].GetAsync();
      string appDisplayName = await this.getServicePrincipal(appId);

      var requestBody = new Permission
      {
        Roles = new List<string>
        {
          role,
        },
        GrantedToIdentities = new List<IdentitySet>
        {
          new IdentitySet
          {
            Application = new Identity
            {
              Id = appId,
              DisplayName = appDisplayName
            },
          },
        },
      };

      try
      {
        await _appGraphClient.Sites[site.Id].Permissions.PostAsync(requestBody);
      }
      catch (Microsoft.Graph.Models.ODataErrors.ODataError ex)
      {
        _logger.LogError(ex.Message);
      }

      return true;
    }

    public async Task<List<AppReg>> GetServicePrincipals(string userAssertion, string prefix)
    {
      _appGraphClient = GetUserGraphClient(userAssertion);
      string filter = $"startswith(DisplayName,'{prefix}')";
      var servicePrincipals = await _appGraphClient.ServicePrincipals.GetAsync(requestConfig =>
      {
        requestConfig.QueryParameters.Filter = filter;
      });
      List<AppReg> appRegs = new List<AppReg>();
      servicePrincipals.Value.ForEach(service =>
      {
        appRegs.Add(new AppReg { Id = service.Id, DisplayName = service.DisplayName });
      });
      return appRegs;
    }

    public async Task<List<SearchSite>> SearchSites(string userAssertion, string queryText)
    {
      _appGraphClient = GetUserGraphClient(userAssertion);
      var sites = await _appGraphClient.Sites.GetAsync((requestConfiguration) =>
      {
        requestConfiguration.QueryParameters.Search = queryText;
      });
      List<SearchSite> siteResults = new List<SearchSite>();
      sites.Value.ForEach(site =>
      {
        siteResults.Add(new SearchSite { Id = site.Id, Title = site.DisplayName, Url = site.WebUrl });
      });
      return siteResults;
    }

    public async Task<bool> IsSiteAdmin(string userAssertion, string siteId, string userMail)
    {
      _appGraphClient = GetUserGraphClient(userAssertion);
      string filter = $"fields/EMail eq '{userMail}'";
      var user = await _appGraphClient.Sites[siteId].Lists["User Information List"].Items.GetAsync(requestConfig =>
      {
        requestConfig.Headers.Add("Prefer", @"HonorNonIndexedQueriesWarningMayFailRandomly");
        requestConfig.QueryParameters.Expand = ["fields"];
        requestConfig.QueryParameters.Select = ["fields"];
        requestConfig.QueryParameters.Filter = filter;
      });
      bool val = user.Value[0].Fields.AdditionalData.TryGetValue("IsSiteAdmin", out object isAdmin);
      if (val)
      {
        return bool.Parse(isAdmin.ToString());
      }
      else
      {
        return false;
      }
    }

    public GraphServiceClient? GetUserGraphClient(string userAssertion)
    {
      var tenantId = _config["tenantId"];
      var clientId = _config["clientId"];
      var clientSecret = _config["clientSecret"];
      var scopes = new[] { "https://graph.microsoft.com/.default" };

      if (string.IsNullOrEmpty(tenantId) ||
          string.IsNullOrEmpty(clientId) ||
          string.IsNullOrEmpty(clientSecret))
      {
        _logger.LogError("Required settings missing: 'tenantId', 'clientId', or 'clientSecret'.");
        return null;
      }

      var onBehalfOfCredential = new OnBehalfOfCredential(
          tenantId, clientId, clientSecret, userAssertion);

      return new GraphServiceClient(onBehalfOfCredential, scopes);
    }

    private async Task<string> getServicePrincipal (string appId)
    {
      string displayName = String.Empty;
      try
      {
        var servicePrincipal = await _appGraphClient.ServicePrincipals[appId].GetAsync();
        displayName = servicePrincipal.DisplayName;
      }
      catch (Microsoft.Graph.Models.ODataErrors.ODataError ex)
      {
        _logger.LogError(ex.Message);
      }

      return displayName;     
  }
}
}
